File: tests/test_tenancy.py
"""Integration tests for tenancy service."""

from fastapi.testclient import TestClient

from tenancy.app import app, service


client = TestClient(app)


def setup_function() -> None:
    service.reset()


def test_first_login_creates_personal_tenant() -> None:
    resp = client.post("/login", headers={"X-User-ID": "user1"})
    assert resp.status_code == 200
    data = resp.json()
    assert data["role"] == "OWNER"
    tenant_id = data["tenant_id"]

    resp2 = client.post("/login", headers={"X-User-ID": "user1"})
    assert resp2.json()["id"] == data["id"]
    assert len(service.tenants) == 1
    assert len(service.memberships) == 1

    resp3 = client.post(
        "/tenants", json={"name": "Team"}, headers={"X-User-ID": "user1"}
    )
    assert resp3.status_code == 200
    assert resp3.json()["name"] == "Team"
    assert len(service.tenants) == 2


def test_invite_and_role_update() -> None:
    owner = client.post("/login", headers={"X-User-ID": "owner"})
    tenant_id = owner.json()["tenant_id"]

    invite = client.post(
        f"/tenants/{tenant_id}/invite",
        json={"user_id": "user2", "role": "MEMBER"},
        headers={"X-User-ID": "owner"},
    )
    assert invite.status_code == 200
    membership_id = invite.json()["id"]

    upd = client.patch(
        f"/memberships/{membership_id}",
        json={"role": "ADMIN"},
        headers={"X-User-ID": "owner"},
    )
    assert upd.status_code == 200
    assert upd.json()["role"] == "ADMIN"

    fail = client.patch(
        f"/memberships/{membership_id}",
        json={"role": "OWNER"},
        headers={"X-User-ID": "user2"},
    )
    assert fail.status_code == 403

File: tests/test_tenancy.go
package tests

import (
    "bytes"
    "encoding/json"
    "net/http"
    "net/http/httptest"
    "testing"

    tenancy "github.com/Cdaprod/ThatDamToolbox/host/services/tenancy"
)

type membership struct {
    ID       string `json:"id"`
    TenantID string `json:"tenant_id"`
    Role     string `json:"role"`
}

type tenant struct {
    ID   string `json:"id"`
    Name string `json:"name"`
}

func TestTenancyFlow(t *testing.T) {
    svc := tenancy.NewService(tenancy.RolePolicy{})
    server := httptest.NewServer(tenancy.NewServer(svc))
    defer server.Close()

    client := server.Client()

    // health
    resp, err := client.Get(server.URL + "/health")
    if err != nil || resp.StatusCode != http.StatusOK {
        t.Fatalf("health failed: %v", err)
    }

    // login
    req, _ := http.NewRequest(http.MethodPost, server.URL+"/login", nil)
    req.Header.Set("X-User-ID", "user1")
    resp, err = client.Do(req)
    if err != nil {
        t.Fatalf("login request failed: %v", err)
    }
    defer resp.Body.Close()
    if resp.StatusCode != http.StatusOK {
        t.Fatalf("login status %d", resp.StatusCode)
    }
    var owner membership
    json.NewDecoder(resp.Body).Decode(&owner)
    if owner.Role != string(tenancy.RoleOwner) {
        t.Fatalf("expected owner role")
    }

    // create tenant
    body, _ := json.Marshal(map[string]string{"name": "Team"})
    req, _ = http.NewRequest(http.MethodPost, server.URL+"/tenants", bytes.NewReader(body))
    req.Header.Set("X-User-ID", "user1")
    resp, err = client.Do(req)
    if err != nil || resp.StatusCode != http.StatusOK {
        t.Fatalf("create tenant failed: %v status %d", err, resp.StatusCode)
    }
    var tnt tenant
    json.NewDecoder(resp.Body).Decode(&tnt)

    // invite
    body, _ = json.Marshal(map[string]string{"user_id": "user2", "role": string(tenancy.RoleMember)})
    req, _ = http.NewRequest(http.MethodPost, server.URL+"/tenants/"+owner.TenantID+"/invite", bytes.NewReader(body))
    req.Header.Set("X-User-ID", "user1")
    resp, err = client.Do(req)
    if err != nil || resp.StatusCode != http.StatusOK {
        t.Fatalf("invite failed: %v status %d", err, resp.StatusCode)
    }
    var member membership
    json.NewDecoder(resp.Body).Decode(&member)

    // membership update
    body, _ = json.Marshal(map[string]string{"role": string(tenancy.RoleAdmin)})
    req, _ = http.NewRequest(http.MethodPatch, server.URL+"/memberships/"+member.ID, bytes.NewReader(body))
    req.Header.Set("X-User-ID", "user1")
    resp, err = client.Do(req)
    if err != nil || resp.StatusCode != http.StatusOK {
        t.Fatalf("update failed: %v status %d", err, resp.StatusCode)
    }
    var updated membership
    json.NewDecoder(resp.Body).Decode(&updated)
    if updated.Role != string(tenancy.RoleAdmin) {
        t.Fatalf("expected role ADMIN, got %s", updated.Role)
    }

    _ = tnt // avoid unused warnings
}

